lib/spack/spack/util/compression.py

```diff
@@ -649,3 +741,44 @@ def check_and_remove_ext(path, ext):
     if check_extension(path, ext):
         return reg_remove_ext(path, ext)
     return path
+
+
+def _substitute_extension(path, old_ext, new_ext):
+    """Return path with old_ext replaced with new_ext.
+    old_ext and new_ext can be extension strings or regexs"""
+    return re.sub(rf"{old_ext}", rf"{new_ext}", path)
+
+
+def expand_contracted_extension_in_path(path, ext=None):
+    """Return path with any contraction extension (i.e. tgz) expanded
+    (i.e. tar.gz). If ext is specified, only attempt to expand that extension"""
+    if not ext:
+        ext = extension_from_path(path)
+    expanded_ext = expand_contracted_extension(ext)
+    if expanded_ext != ext:
+        return _substitute_extension(path, ext, expanded_ext)
+    return path
+
+
+def expand_contracted_extension(extension):
+    """Return expanded version of contracted extension
+    i.e. .tgz -> .tar.gz, no op on non contracted extensions"""
+    extension = extension.strip(".")
+    if extension in NOTAR_EXTS and extension != "zip":
+        # NOTAR exts are abbreviated extensions
+        # for compressed tarballs
+        # so we know tar is the first ext after initial '.'
+        prefix = "tar"
+        suffix = extension[1:]
```

Comment:
If extension is a string like "tar", it looks like this would be problematic (part of why I asked about preceding "." above was because I couldn't think of another reason for stripping off the first item of extension). Could you also explicitly say in the docstring that extension is a string (i.e. "tar.gz" rather than a list like `["tar", "gz"]`

Reply:
I'm always happy to improve a docstring, will do. 
With regard to the tar concern, maybe we should discuss at the meeting tomorrow, as I'm a bit unsure how this could be problematic. For example, I just tried this out locally:
```[spack] C:\spack_win\spack>spack python
Spack version 0.20.0.dev0
Python 3.8.4, Windows AMD64
>>> import spack.util.compression as sc
>>> sc.expand_contracted_extension("tar")
'tar'
>>> sc.expand_contracted_extension(".tar")
'tar'
>>> sc.expand_contracted_extension(".tar.gz")
'tar.gz'
>>> sc.expand_contracted_extension(".tgz")
'tar.gz'
>>> sc.expand_contracted_extension(".tar.bz2")
'tar.bz2'
>>> sc.expand_contracted_extension(".tbz2")
'tar.bz2'
>>> sc.expand_contracted_extension(".tbz")
'tar.bz2'
```
Which seems to be behaving as expected.

Reply:
I also don't believe an extension "tar" would make it to that code block as the preceding if statement would preclude that. At that point in execution, we would know that the "extension" is one of `"tgz", "txz", "tbz2", "tbz"` so stripping off the first character (the t) would give us the true extension (with the exception of `tbz` which I handle later).

Reply:
OK I see that all `NOTAR_EXTS` except zip/tbz would be handled this way. In my case I didn't understand how this worked until you pointed that out. IMO a dictionary like {"tbz": "bz2", "tgz": "tar.gz"...} would be more straightforward (and also possibly avoid future issues if more entries are added to `NOTAR_EXTS`).

Reply:
I've been avoiding taking that approach as I assumed that a hardcoded dictionary supplementary to `NOTAR_EXTS` would just be an added and easy to miss maintenance burden. If not, I can add it.